### PR TITLE
Introduce soft assertions for MockMvc

### PR DESCRIPTION
**Pull request by wyhasany**
It happens very often that MockMvc is used in
heavyweight integration tests. It's no use to waste
time to check if another condition has been fixed or
not. Soft assertion could help a lot to check all
conditions at once even if one of them fail.

### CODE REVIEW

**Comment by sbrannen** [comment]
Related to #20963

File: spring-test/src/main/java/org/springframework/test/web/servlet/ResultMatcher.java [hunk 1]
```diff
@@ -70,4 +73,32 @@ static ResultMatcher matchAll(ResultMatcher... matchers) {
...
+				catch (AssertionError assertionException) {
+					if (failedMessages == null) {
+						failedMessages = new ArrayList<>();
+					}
+					failedMessages.add(assertionException.getMessage());
+				}
+			}
+			if (failedMessages != null) {
+				throw new AssertionError(String.join("\n", failedMessages));
+			}
```
Comment (nguyensach) [inline spring-test/src/main/java/org/springframework/test/web/servlet/ResultMatcher.java:100]:
```suggestion
			List<String> failedMessages = new ArrayList<>();
			for (ResultMatcher matcher : matchers) {
				try {
					matcher.match(result);
				}
				catch (AssertionError assertionException) {
					failedMessages.add(assertionException.getMessage());
				}
			}
			if (!failedMessages.isEmpty()) {
				throw new AssertionError(String.join("\n", failedMessages));
			}
```

**Comment by wyhasany** [comment]
@nguyensach I've applied your suggestions. Do you need any action on my side?

**Review by nguyensach — APPROVED** [review]
LGTM 🚀

File: spring-test/src/main/java/org/springframework/test/web/servlet/ResultMatcher.java [hunk 2]
```diff
@@ -70,4 +73,29 @@ static ResultMatcher matchAll(ResultMatcher... matchers) {
...
+	 */
+	static ResultMatcher matchAllSoftly(ResultMatcher... matchers) {
+		return result -> {
+			List<String> failedMessages = new ArrayList<>();
+			for (ResultMatcher matcher : matchers) {
+				try {
+					matcher.match(result);
+				}
+				catch (AssertionError assertionException) {
+					failedMessages.add(assertionException.getMessage());
```
Comment (kkocel) [inline spring-test/src/main/java/org/springframework/test/web/servlet/ResultMatcher.java:92]: to be consistent with WebTest version you could consider adding indices here https://github.com/spring-projects/spring-framework/pull/26969/files#diff-396f20e4c7198aab834e4d234e312f65e9fedf27d32582f336d1e8b0ca0a0328R521

**Review by kkocel — CHANGES_REQUESTED** [review]

File: spring-test/src/main/java/org/springframework/test/web/servlet/ResultMatcher.java (hunk 2 above)
Comment (wyhasany) [reply spring-test/src/main/java/org/springframework/test/web/servlet/ResultMatcher.java]: good idea, fixed

**Comment by sbrannen** [comment]
Current work on this issue can be seen in the following feature branch: https://github.com/spring-projects/spring-framework/compare/main...sbrannen:issues/gh-26917-MockMvc-soft-assertions

**Comment by sbrannen** [comment]
Thanks for the PR, @wyhasany! 👍 

This has been merged into `main` in 35bec8102b014feac18c5b096b39c3592c27b76e, refined in 5f47d3be227e115c57b03e73f5df97993c44fdd8 and cd078eaad83e2c6cb441719086b460e665c20f35, and then fully revised in dd9b99e13d6b30462fb168799e6bbca1456123da.

In the end we decided to introduce a new `andExpectAll()` method in `ResultActions` and to move away from static methods in `ResultMatcher`. In fact, we even deprecated the existing `matchAll()` method in `ResultMatcher`.

**Pull request closed without merging**